mlserver/batching/requests.py

```diff
@@ -0,0 +1,121 @@
+from operator import mul
+from collections import defaultdict
+from typing import Any, Dict, Iterable, List, Union
+from functools import reduce
+
+from ..types import (
+    InferenceRequest,
+    InferenceResponse,
+    RequestInput,
+    ResponseOutput,
+)
+
+
+def _get_data(payload: Union[RequestInput, ResponseOutput]):
+    return getattr(payload.data, "__root__", payload.data)
+
+
+def _infer_elem_size(shape: List[int]) -> int:
+    # TODO: Allow to use a different batch dimension
+    return reduce(mul, shape[1:], 1)
+
+
+def _merge_data(request_inputs: List[RequestInput]) -> Any:
+    all_data = [_get_data(request_input) for request_input in request_inputs]
+
+    sampled_datum = all_data[0]
+
+    if isinstance(sampled_datum, str):
+        return "".join(all_data)
+
+    if isinstance(sampled_datum, bytes):
+        return b"".join(all_data)
+
+    if isinstance(sampled_datum, list):
+        return sum(all_data, [])
+
+    # TODO: Should we raise an error if we couldn't merge the data?
+    return all_data
+
+
+def _split_data(response_output: ResponseOutput) -> Any:
+    element_size = _infer_elem_size(response_output.shape)
+    merged_data = _get_data(response_output)
+
+    # TODO: Don't rely on array to have been flattened
+    for i in range(0, len(merged_data), element_size):
+        yield merged_data[i : i + element_size]
+
+
+class BatchedRequests:
+    def __init__(self, inference_requests: List[InferenceRequest] = []):
+        self._inference_requests = inference_requests
+        self.merged_request = self._merge_requests()
+        self._prediction_ids: List[str] = [  # type: ignore
+            req.id for req in self._inference_requests  # type: ignore
+        ]
+
+    def _merge_requests(self) -> InferenceRequest:
+        inputs_index: Dict[str, List[RequestInput]] = defaultdict(list)
+
+        for inference_request in self._inference_requests:
+            # TODO: What should happen if UID is empty?
+            for request_input in inference_request.inputs:
+                inputs_index[request_input.name].append(request_input)
+
+        # TODO: What should happen if input cardinality is different? (e.g.
+        # inputs with 2 inputs and others with 1)
```

Comment:
Good point, I think we could put a constraint that it should always have the same number of inputs, or have a default... but the latter could be harder - I can't think of a situation where a model could take an input as optional (unless it's some custom python code). Maybe if you just add it with an error and pick it up when a usecase comes up

Reply:
Definitely an error I'd say

Reply:
Just to clarify, my impression is that there are two different axes here. On one hand, we could have different inputs with a different number of "input heads". This is the tricky case, involving optional parameters, where we should totally raise an error. 

On the other hand, we have the case where a payload could already have multiple entries within the same input head (i.e. the request has a "mini batch" within it already). In this case, we could just append the `n` elements contained within that request, and just keep the count of how many elements correspond to each request ID. 

Before adding that in though, it would be great to hear your thoughts about the latter case.

Reply:
Is there an example of the former case?

For the latter it should certainly add to mini-batches together I think. But if hard could be a later change.

Reply:
The latter (i.e. merging variable size mini-batches) should now be also supported :+1: